### ISSUE TITLE
ANN401 doesn't trigger for generic arguments

### ISSUE THREAD

**Issue by ItsDrike** [astral-sh/ruff#5803]
From the following code snippet:
```python
from __future__ import annotations

from typing import Any, Generic, TypeVar

T = TypeVar("T")

a: Any = 5
b: dict[str, Any] = {}
c: list[Any] = []


class Foo:
    X: Any = 5
    Y: dict[str, Any] = {}


class Bar(Generic[T]):
    ...


x: Bar[Any] = Bar()


def foo(x: Any) -> None:
    ...


def bar(x: dict[str, Any]) -> None:
    ...
```

When running `ruff check --isolated --fix --show-source --show-fixes --select ANN401 test.py`, ruff only reports 1 ANN401 violation:
```
test.py:22:12: ANN401 Dynamically typed expressions (typing.Any) are disallowed in `x`
   |
22 | def foo(x: Any) -> None:
   |            ^^^ ANN401
23 |     ...
   |
```

However `flake8-annotations` would treat every use of `Any` as a violation here. It seems that ruff's implementation of ANN401 doesn't trigger when Any is used as a generic parameter somewhere, it only triggers when used directly.

Ideally, ANN401 should be changed to catch every use of Any, including the use in generic arguments, or there should be a new rule that can do that. (However considering `flake8-annotations` would catch these with ANN401, ruff should probably do the same)

**Comment by dhruvmanila** [comment]
Hey, thanks for using Ruff and opening this issue.

> Any in generic arguments

You've mentioned here that `flake8-annotations` raises a violation for the provided example  in this section but I'm unable to reproduce this. Can you provide the command you used to run `flake8` on the example? I've used the following command:

```console
flake8 --extend-select=ANN401 test.py
```

It's also mentioned in the plugin README that this rule isn't supported for nested dynamic types (https://github.com/sco1/flake8-annotations#dynamic-typing-caveats):

> Nested dynamic types (e.g. `typing.Tuple[typing.Any]`) and redefinition (e.g. `from typing import Any as Foo`) will not be identified.

_Ruff identifies redefinition_

---

> It seems that ANN401 only triggers when used in a function annotation, variables are not checked.

Yes, this is true as is the case for `flake8-annotations`:

> Only function declarations are considered by this plugin; type annotations in function/module bodies are not checked

---

> Any in concrete overload

This seems like a bug as it should be raising violations similar to the plugin. Thanks for noticing this!

---

Overall, I think we should improve the documentation to be clear on where this rule will be checked and enable it on `@overload` implementation function.

**Comment by charliermarsh** [comment]
(Omitting Any in a concrete overload might be desirable, is that not a common pattern?)

**Comment by dhruvmanila** [comment]
I'm not sure if it's a common pattern as the concrete annotation would probably be a union of all possible types. Although, `flake8-annotations` does raise violation in this case.

**Comment by ItsDrike** [comment]
Oh, you're actually right, it seems that `flake8-annotations` also doesn't check this, huh, I've only noticed this when moving our code-base to ruff, as Ruff has reported unused noqa comments. Apparently this was never an issue, and those comments were just added because someone assumed it would be.

That's pretty interesting, so in that case, I'd like to make this a suggestion, as I think there is value in having a linter rule capable of finding use of `Any` in those nested type definitions, and it would be really nice if we were able to detect variable annotations containing `Any` too.

As for the overload, after thinking on it for a bit, yeah, it should probably not behave like this, even though to the outside user of that function, the `Any` type there would not be visible, it should still absolutely be a union of those types, or a supertype shared by all types in the overloads, to allow the type-checker to properly check the implementation of that function itself. So that is a bug which should get fixed.

**Comment by dhruvmanila** [comment]
> That's pretty interesting, so in that case, I'd like to make this a feature request, as I think there is value in having a linter rule capable of finding use of `Any` in those nested type definitions, and it would be really nice if we were able to detect variable annotations containing `Any` too.

If possible can you open a new issue for this (better for tracking)? Otherwise I can open it later as well :)

> As for the overload, after thinking on it for a bit, yeah, it should probably not behave like this, even though to the outside user of that function, the `Any` type there would not be visible, it should still absolutely be a union of those types, or a supertype shared by all types in the overloads, to allow the type-checker to properly check the code of the actual implementation of that function. So that is a bug which should get fixed.

I agree. Any thoughts here? /cc @charliermarsh @zanieb

**Comment by ItsDrike** [comment]
Alright, I've opened a separate issue for the feature request, leaving this one just as a bug report.

By the way, I've also noticed another behavior that Ruff doesn't pick up on, this one might be harder to address though, and I would understand if it was not something you'll want to fix:

```python
from typing import TypeAlias, Any

MyAny: TypeAlias = Any

def foo(x: MyAny) -> None:
    ...
```

In this code, Ruff doesn't report any annotation issues, even though `x` is actually annotated with the `Any` type, it's just hidden behind a type alias.

**Comment by dhruvmanila** [comment]
> Alright, I've opened a separate issue for the feature request, leaving this one just as a bug report.

Thanks!

> In this code, Ruff doesn't report any annotation issues, even though `x` is actually annotated with the `Any` type, it's just hidden behind a type alias.

Yes, type aliases are a known issue across multiple rules as they're difficult to detect reliably.

**Comment by ItsDrike** [comment]
> Yes, type aliases are a known issue across multiple rules as they're difficult to detect reliably.

That's totally fair, though it should probably be mentioned in the rule's docs, to make this clear.
